pkg/lib/synthesize/generate.go

```diff
@@ -0,0 +1,135 @@
+)
+
+const SynthesizedCSVFileName string = "olm.clusterserviceversion.yaml"
+
+// Generate generates a CSV for the bundle based on the resource within the bundle directory.
+// The bundleDir points to where the bundle is located, outputDir is where bundle and synthesized CSV will be written.
+// If overwrite is set to true, outputDir is reloaded with the exact content of bundleDir plus a synthesized CSV
+// and inplace generation deletes the CSV if exists.
+func Generate(bundleDir, outputDir string, overwrite bool) error {
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	if outputDir != "" {
+		outputDir, err = filepath.Abs(outputDir)
+		if err != nil {
+			return err
+		}
+	} else {
+		// if outputDir not specified, synthesis within the bundle directory.
+		outputDir = bundleDir
+	}
+
+	bundle, err := SynthesizeFromDirectory(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	csv, err := bundle.GenerateCSV()
+	if err != nil {
+		return err
+	}
+
+	if bundle.existingCSVPath != "" && !overwrite {
+		return fmt.Errorf("a ClusterServiceVersion already exist in the bundle at %s", bundle.existingCSVPath)
+	}
+
+	// Only delete (overwrite) the existing CSV in a bundle if specified to overwrite for inplace generation.
+	if bundle.existingCSVPath != "" && outputDir == bundleDir {
+		if err := os.Remove(bundle.existingCSVPath); err != nil {
+			return fmt.Errorf("error overwriting ClusterServiceVersion %s, %v", bundle.existingCSVPath, err)
+		}
+	} else if outputDir != bundleDir {
+		// Based on the assumption that CSV is only useful to reside in a bundle that has the exact resource,
+		// this overwrites the entire output directory so that is has the exact bundle manifests as the bundledir.
+		if overwrite {
+			// Do not expect path error since path is examined earlier.
+			os.RemoveAll(outputDir)
+		}
+		if err := CopyAllFilesFromDirectoryExcept(bundleDir, outputDir, []string{bundle.existingCSVPath}); err != nil {
+			return err
+		}
+	}
+
+	// generate CSV file to the outputDir.
+	return WriteStructToFile(csv, csv.GroupVersionKind(), SynthesizedCSVFileName, outputDir)
+}
+
+// WriteStructToFile writes a struct's content to a YAML file.
+func WriteStructToFile(obj runtime.Object, gvk schema.GroupVersionKind, fileName, outputPath string) error {
+	serial := serializer.NewSerializerWithOptions(nil, nil, nil, serializer.SerializerOptions{true, false, false})
+
+	if _, err := os.Stat(outputPath); os.IsNotExist(err) {
+		err = os.MkdirAll(outputPath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	filePath := path.Join(outputPath, fileName)
+	f, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+
+	if err := serial.Encode(obj, f); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CopyAllFilesFromDirectoryExcept recursively copy files from one directory to another and truncates a file if exists.
+func CopyAllFilesFromDirectoryExcept(srcPath, dstPath string, files []string) error {
```

Comment:
The name of this function is a little confusing to me. Seems like it's trying to copy all files from a directory `Except` one file, and I go looking for which file it's trying to omit.